### PR TITLE
feature: add optics-xapps typechain gen

### LINKED ISSUE #375: add typechain generation to optics-xapps
follow pattern in prestwich/ts-deploy and #374

### PR DESCRIPTION
**Pull request by YoduYodu**
### Description

-  Add typechain generation to optics-xapps.
- Reorg typescript/src/typechain:
```
- typechain
  - optics-core
  - optics-xapps
  - [more contracts]
```

Fixes #375

### CODE REVIEW

File: solidity/optics-xapps/hardhat.config.js [hunk 1]
```diff
@@ -25,4 +25,10 @@ module.exports = {
       url: 'http://localhost:8545',
     },
   },
+
+  typechain: {
+    outDir: '../../typescript/src/typechain',
```
Comment (prestwich) [inline solidity/optics-xapps/hardhat.config.js:30]: I think this will overwrite the core contracts. instead, provide a separate directory for them

**Review by prestwich — CHANGES_REQUESTED** [review]

File: solidity/optics-xapps/hardhat.config.js (hunk 1 above)
Comment (YoduYodu) [reply solidity/optics-xapps/hardhat.config.js]: I thought about that as well. So I've gone ahead and reorged it as follows
- typechain
    - optics-core
    - optics-xapps
    - [more contracts]

**Comment by ErinHales** [comment]
You'll need to update the imports for the files you moved from '/typechain' to '/typechain/optics-xapps'.

**Comment by ErinHales** [comment]
And, as we discussed, next step will be to add the typechain gen to the 'npm run compile' script and the pre-commit script

**Comment by prestwich** [comment]
@YoduYodu it looks like you ran into the `git v2.24` bug and it deleted all files in the repo. you'll need to update your git version `brew install git` and then reload your terminal. you'll also want to check out the previous commit `git checkout main~1` and then redo whatever work was lost

@ErinHales any idea why the precommit script didn't stop him from using git v2.24?

**Comment by ErinHales** [comment]
Oh no 😬 I will look into it

**Review by ErinHales — APPROVED** [review]
Looks good!

**Pull request merged**